src/biotite/sequence/profile.py

```diff
@@ -0,0 +1,165 @@
+# under the 3-Clause BSD License. Please see 'LICENSE.rst' for further
+# information.
+
+import numpy as np
+from .seqtypes import NucleotideSequence, ProteinSequence
+from .alphabet import LetterAlphabet
+
+__name__ = "biotite.sequence"
+__author__ = "Maximilian Greil"
+_all__ = ["SequenceProfile", "to_consensus"]
+
+# Abbreviations
+_NUC_DNA_ALPH = NucleotideSequence.alphabet_unamb
+_NUC_RNA_ALPH = LetterAlphabet(["A", "C", "G", "U"])
+_PROT_ALPH = ProteinSequence.alphabet
+
+
+class SequenceProfile(object):
+    """
+    A :class:`SequenceProfile` object stores information about a sequence
+    profile of aligned sequences, i.e. the consensus sequence.
+
+    This class saves the position frequency matrix of the occurrences of
+    each alphabet symbol at each position. It also saves the number of gaps
+    at each position.
+
+    A :class:`SequenceProfile` object can be created from an indefinite number
+    of aligned sequences.
+
+    All attributes of this class are publicly accessible.
+
+    Parameters
+    ----------
+    position_frequency_matrix : ndarray, dtype=int, shape=(n,k)
+        This matrix simply saves for each position how often absolutely each symbol is present.
+    position_gaps : ndarray, dtype=int, shape=n
+        Array which indicates the number of gaps at each position.
+    alphabet : Alphabet, length=k
+        Alphabet of sequences of sequence profile
+
+    Attributes
+    ----------
+    position_frequency_matrix : ndarray, dtype=int, shape=(n,k)
+        This matrix simply saves for each position how often absolutely each symbol is present.
+    position_gaps : ndarray, dtype=int, shape=n
+        Array which indicates the number of gaps at each position.
+    alphabet : Alphabet, length=k
+        Alphabet of sequences of sequence profile
+    """
+
+    def __init__(self, position_frequency_matrix, position_gaps, alphabet):
+        self.position_frequency_matrix = position_frequency_matrix
+        self.position_gaps = position_gaps
+        self.alphabet = alphabet
+
+    def __repr__(self):
+        """Represent SequenceProfile as a string for debugging."""
+        return f"SequenceProfile(np.{np.array_repr(self.position_frequency_matrix)}, " \
+               f"np.{np.array_repr(self.position_gaps)}, Alphabet({self.alphabet}))"
+
+    def __eq__(self, item):
+        if not isinstance(item, SequenceProfile):
+            return False
+        if not np.array_equal(self.position_frequency_matrix, item.position_frequency_matrix):
+            return False
+        if not np.array_equal(self.position_gaps, item.position_gaps):
+            return False
+        if not self.alphabet == item.alphabet:
+            return False
+        return True
+
+    @staticmethod
+    def from_alignment(alignment):
+        """
+        Get an object of :class:`SequenceProfile` from an object of :class:`Alignment`
+        """
+        sequences = alignment.get_gapped_sequences()
```

Comment:
Using `get_gapped_sequences()` in this place is quite inefficient: The trace is converted to symbol codes, the symbol codes to symbols and the symbols into a string, and vice versa.

Probably `sequence.align.get_codes()` (https://www.biotite-python.org/apidoc/biotite.sequence.align.get_codes.html) is more efficient in this place. Then you can use `np.bincount()`for every column. You could handle `-1` values by replacing them with `len(alphabet)`